affinity/cpp-20/d0796r1.md

```diff
@@ -95,395 +96,423 @@ Some systems give additional user control through explicit binding of threads to
 
-    /* Execution resource */
+Two important considerations when defining a unified interface for querying the **resource topology** of a **system**, are (a) what level of abstraction such an interface should have, and (b) at what granularity it should describe the topology's **execution resources**. As both the level of abstraction of an **execution resource** and the granularity that it is described in will vary greatly from one implementation to another, it’s important for the interface to be generic enough to support any level of abstraction. To achieve this we propose a generic hierarchical structure of **execution resources**, each **execution resource** being composed of other **execution resources** recursively. Each **execution resource** within this hierarchy can be used to place memory (i.e., allocate memory within the **execution resource’s** memory region), place execution (i.e. bind an execution to an **execution resource’s execution agents**), or both.
 
-    struct execution_resource {
+> For example, a NUMA system will likely have a hierarchy of nodes, each capable of placing memory and placing agents.  A CPU + GPU system may have GPU local memory regions capable of placing memory, but not capable of placing agents.
 
-      execution_resource() = delete;
-      execution_resource(const execution_resource &) = delete;
-      execution_resource(execution_resource &&) = delete;
-      execution_resource &operator=(const execution_resource &) = delete;
-      execution_resource &operator=(execution_resource &&) = delete;
-      ~execution_resource();
+Nowadays, there are various APIs and libraries that enable this functionality. One of the most commonly used is the Portable Hardware Locality (hwloc) [9]. Hwloc presents the hardware as a tree, where the root node represents the whole machine and subsequent levels represent different partitions depending on different hardware characteristics. The picture below shows the output of the hwloc visualization tool (lstopo) on a 2-socket Xeon E5300 server. Note that each socket is represented by a package in the graph. Each socket contains its own cache memories, but both share the same NUMA memory region. Note also that different I/O units are visible underneath: Placement of these units with respect to memory and threads can be critical to performance. The ability to place threads and/or allocate memory appropriately on the different components of this system is an important part of the process of application development, especially as hardware architectures get more complex. The documentation of lstopo [22] shows more interesting examples of topologies that appear on today's systems.
```

Comment:
"One of the most commonly used is the Portable Hardware Locality (hwloc) library [9]."

Reply:
Fixed.